### PR TITLE
MultiOperation: write to cache before persist

### PR DESCRIPTION
**Pull request by stephanos**
## What changed?

Moved the cache write before the persistence write.

## Why?

The code has a comment to explain it.

## How did you test it?

I have a test runner that concurrently executes MultiOperations.

### CODE REVIEW

File: service/history/api/updateworkflow/api.go [hunk 1]
```diff
@@ -125,9 +129,9 @@ func (u *Updater) Invoke(
 
 func (u *Updater) ApplyRequest(
 	ctx context.Context,
-	workflowLease api.WorkflowLease,
+	updateReg update.Registry,
+	ms workflow.MutableState,
```
Comment (stephanos) [inline service/history/api/updateworkflow/api.go:133]: This interface makes it much clearer what's needed for this function.

File: service/history/api/multioperation/api.go [hunk 2]
```diff
@@ -248,23 +268,10 @@ func startAndUpdateWorkflow(
...
-		shardContext.GetMetricsHandler(),
-	); err != nil {
-		return nil, err
-	}
-
 	// wait for the update to complete
 	updateResp, err := updater.OnSuccess(ctx)
 	if err != nil {
-		return nil, serviceerror.NewUnavailable(fmt.Errorf("failed to complete Workflow Update: %w", err).Error())
+		return nil, newMultiOpError(nil, err) // `nil` for start since it succeeded
```
Comment (stephanos) [inline service/history/api/multioperation/api.go:274]: Still not 100% sure what this should be; but this seems better than before. Maybe.

File: service/history/api/updateworkflow/api.go (hunk 1 above)
Comment (stephanos) [reply service/history/api/updateworkflow/api.go:133]: However, I'd like to see a world in the future where I only need to define a single parameter here.

File: service/history/api/multioperation/api.go (hunk 2 above)
Comment (alexshtin) [reply service/history/api/multioperation/api.go:274]: It looks definitely better to me.

**Review by alexshtin — APPROVED** [review]

**Pull request merged**